### PR TITLE
[WIP] Connectionist temporal classification loss example

### PR DESCRIPTION
**Pull request by duvenaud**
The connectionist temporal classification (CTC) loss should be a good demo for Dex, since it's a partially-parallel dynamic programming algorithm.  This implementation ended up being very short, because it (if autodiff can handle it) doesn't need the backward pass explicitly for the gradient.  Overall I was pretty happy with how closely the implementation ended up looking like the equations from the paper.  I think Dex might also be nice for writing other sequence-handling algorithms like Beam search.

I was especially happy that I could express the type of "prepend and interleave" which turns "text" into " t e x t ".  Although it had a complicated enough type that I ended up contorting my code a bit to avoid spelling it in the main loop.  I might have had use for some sort of ```typeof``` function.

The 'last mile' of making the code look like pseudocode / math could be approached by:

1) (Suggested by Adam) Creating a 'logReal' type, which overloads (+, *) to use (logaddexp, +).  I'm game to try this but don't know the syntax.
2) Automatically putting implicit binders into scope, to avoid a bunch of implicit arguments to the main function.  It might also make sense to automatically put typeclass dictionaries into scope (like Eq), without explicitly adding the implicit annotation.  I ended up needing 3 of those for the ctc function type, which ended up having 6 implicit binders.
3) Letting the user create custom index sets more easily, perhaps specifying a function that handles their indexing.  This would let me hide my index-munging code in the main loop, although it might hurt readability.
4) I had to use ```select``` more than I would have liked, because it was hard to prepend to the beginning of a vector.  Also, the syntax for starting a for loop not at the first element, e.g. ```for i:((1@)..). i``` appears to be broken.  This also made it hard to sum over the last two elements of the last time step.

Warning: I don't really know how to test this code, and suspect that more likely than not it has a bug.

For contrast, here are some other (more fully-featured) implementations:
1. [Tensorflow](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/ops/ctc_ops.py#L827-L929)
2. [Julia](https://github.com/FluxML/Flux.jl/pull/342/files)
3. [C++ / CUDA](https://github.com/baidu-research/warp-ctc/blob/6d5b8fac130638862d97dc48ef43a8d7b5a503bb/include/detail/gpu_ctc_kernels.h#L350-L360)

### CODE REVIEW

File: examples/ctc.dx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  m = max x y
+  m + ( log ( (exp (x - m) + exp (y - m))))
+
+def ctc (time : Type) ?-> (vocab : Type) ?-> (position : Type) ?->
+        (dict: Eq vocab) ?=> (dict2: Eq position) ?=> (dict3: Eq time) ?=> (blank: vocab)
+        (logits: time=>vocab=>Real) (labels: position=>vocab) : Real =
+  -- Computes log p(labels | logits), marginalizing over possible alignments.
+  -- Todo: remove unnecessary implicit type annotations once
+  -- Dex starts putting implicit types in scope.
+
```
Comment (dougalm) [inline examples/ctc.dx:57]: Yes, I agree that putting the implicit implicit arguments in scope will reduce a lot of clutter. Making the dictionaries implicit will be harder but perhaps we can at least make it syntactically lighter. Even now, you can use `_` instead of names like `dict`.

File: examples/ctc.dx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    select (t == (0@_)) log_prob_seq_t0 for s.
+      cond = ilabels.s == blank || same_as_last ilabels s
+      labar = logaddexp prev.s (safe_idx prev ((ordinal s) - 1))
+      other = logaddexp labar  (safe_idx prev ((ordinal s) - 2))
+      ans = select cond labar other
+      ans + normalized_logits.t.(ilabels.s)
+
+  log_prob_seq_final = fold log_prob_seq_t0 update
+
+  -- Todo: nicer way to get last two elements of log_prob_seq_final.
```
Comment (dougalm) [inline examples/ctc.dx:87]: We should probably just add functions like `firstIdx : Ix n ?=> n` and `lastIdx : Ix n ?=> n`. Unfortunately, these are partial functions because the index sets might be empty.

File: examples/ctc.dx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+most implementations.
+
+
+def interleave (blank:v) (labels: m=>v) : (m & (Fin 2))=>v =
+  -- Turns "text" into "t e x t " by first pairing each letter with a blank,
+  -- then flattening the pairs back into a single-index table.
+  pairs = for i. [labels.i, blank]
+  for (i, j). pairs.i.j
+
+def prepend (first: v) (seq: m=>v) : (Unit|m)=>v =
```
Comment (dougalm) [inline examples/ctc.dx:27]: It's cool that this sum-type index set worked! I think this is the first example of it being used in anger.

File: examples/ctc.dx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  -- Turns "text" into " t e x t".
+  -- The output of this function has a slightly complicated output type, which
+  -- has size 1 + 2 * (size m).
+  interleaved = interleave blank seq
+  prepend blank interleaved
+
+def clipidx (n:Type) -> (i:Int) : n =
+  -- Returns element at 0 if less than zero.
+  -- Ideally we could have an alternative
+  -- to Fin that just clips the index at its bounds.
```
[the remark this answers is not in the document]
Comment (duvenaud) [reply examples/ctc.dx:44]: Agreed, although this algorithm also needs to double-decrement, which makes this approach less appealing in this instance.

File: examples/ctc.dx (hunk 1 above)
Comment (duvenaud) [reply examples/ctc.dx:57]: Ok.  Also, if I can get rid of some of the ```select i == j``` then I think I need only 1 Eq dictionary.

File: examples/ctc.dx (hunk 2 above)
Comment (duvenaud) [reply examples/ctc.dx:87]: We might be able to address this issue with more typeclasses, e.g. ```firstIdx``` and ```lastIdx``` might be entries of a ```nonEmptyIx``` dictionary.

File: examples/ctc.dx (hunk 3 above)
Comment (duvenaud) [reply examples/ctc.dx:27]: I liked that the index set can be both iterated over in a flat manner, but that the type of the index set "remembers" where the different elements came from.

File: examples/ctc.dx (hunk 4 above)
[the remark this answers is not in the document]
Comment (dougalm) [reply examples/ctc.dx:44]: With the above, you can write `i -| 2`. Maybe you mean something else by "double-decrement"?

[the remark this answers is not in the document]
Comment (duvenaud) [reply examples/ctc.dx:44]: D'oh, I didn't read carefully enough!  Yes, that'd work.

The other thing that'd make this demo nicer is if one could specify a value to return when indexing goes out of bounds.

[the remark this answers is not in the document]
Comment (apaszke) [reply examples/ctc.dx:44]: The out-of-bounds case would be really easy to handle if only we had a dual to `inject` (called `project`?):
```hs
project : (n : Type) ?-> (x : Int) : Maybe n = ...
```
Then, you could do "infinite padding" on an array by simply wrapping it in a function:
```hs
pad : (arr : n=>a) (elem : a) : Int->a = \i.
  case project i of
    Just idx -> arr.idx
    Nothing  -> a
```

File: examples/ctc.dx (hunk 2 above)
Comment (apaszke) [reply examples/ctc.dx:87]: Yeah, but it would be hard to prove that a type really should have `nonEmptyIx` statically. Certainly `Fin n`, for a run-time computed `n` doesn't have it, and neither do most range types.

File: examples/ctc.dx (hunk 4 above)
[the remark this answers is not in the document]
Comment (duvenaud) [reply examples/ctc.dx:44]: @apaszke  I like that idea.  One thing I want to mention is that I only needed a ```clipdix``` function here because ```select``` doesn't short circuit.  So even though I had switching logic to avoid out-of-bounds indexing, because both sides were evaluated, the unused branch would cause a runtime error unless I wrapped it in ```clipdix```.

File: examples/ctc.dx (hunk 2 above)
Comment (duvenaud) [reply examples/ctc.dx:87]: Maybe, but I'd love to go down this road a little.  It does have a Turing-tarpit flavor, but now that Dex has dependent types I wonder if most of these guarantees couldn't be passed around mostly automatically.  For instance, in this demo, the padded text already has a complicated type that does guarantee it's non-empty.  But pleasingly, I can still iterate over it without spelling this complicated type.  So maybe these extra proofs / guarantees could ride along with the type implicitly in most user code.

Comment (apaszke) [reply examples/ctc.dx:87]: We could certainly expand the dictionary-synthesis code to do some more elaborate proof search, but it will not be able to handle all possible use cases. E.g. `Fin n` for `n = 1 + abs x` is always non-empty, but it is quite a non-trivial proof that requires reasoning about integer arithmetic which is very hard to do in general.
